data/completion/complete.sh

```diff
@@ -0,0 +1,73 @@
+# -*- bash -*-
+
+# _complete_from_snap serialises the tab completion request and sends it off to
+# the appropriate 'snap run --command=complete', and de-serialises the response
+# into the usual tab completion result.
+_complete_from_snap() {
+    {
+        read -a opts
+        # opts is expected to be a series of compopt options
+        if [[ ${#opts[@]} -gt 0 ]]; then
+            if [[ "${opts[0]}" == "cannot" ]]; then
+                # older snap-execs sent errors over stdout :-(
+                return 1
+            fi
+
+            for i in "${opts[@]}"; do
+                if ! [[ "$i" =~ ^[a-z]+$ ]]; then
+                    # non-alphanumeric option; something awry
+                    return 2
+                fi
+            done
+        fi
+
+        read bounced
+        case "$bounced" in
+            ""|"alias"|"export"|"job"|"variable")
+                # OK
+                ;;
+            *)
+                # unrecognised bounce
+                return 2
+                ;;
+        esac
+
+        read sep
+        if [ "$sep" ]; then
+            # non-blank separator? madness!
+            return 2
+        fi
+        local oldIFS="$IFS"
+
+        if [ ! "$bounced" ]; then
+            local IFS=$'\n'
+            COMPREPLY=( $( \grep -v '[[:cntrl:];?*{}]' ) )
+            IFS="$oldIFS"
+        fi
+
+        if [[ ${#opts[@]} -gt 0 ]]; then
+            compopt $(printf " -o %s" "${opts[@]}")
+        fi
+        if [ "$bounced" ]; then
+            COMPREPLY+=(compgen -A "$bounced" -- "${COMP_WORDS[$COMP_CWORD]}")
```

Comment:
Ok, so if I change complete.sh to have (there are better ways to do this):
```
            local reply=( $( cat ) )
            for i in $(seq 0 $((${#reply[@]} - 1))); do
                COMPREPLY[$i]=`printf "%q" "${reply[$i]}"`
            done
```

Then get:
```
$ mkdir foo ; cd foo
$ touch "fooz'ball" 'party "animal"' "disc (1)" "norf:qux"
$ complexion <tab>
  |
   -> $ complexion
      disc\ \(1\)        fooz\'ball         norf\;qux          party\ \"animal\"
      $ complexion n<tab>
        |
         -> $ complexion norf\;qux
            Greetings from inside /snap/complexion/x3.
            Arguments given:
            > norf\;qux
            That is all. Have a nice day.
```

This at least makes it so the actual completion is both safe and usable. However, it still differs from 'ls' output. It also isn't likely going to work right for command completion that might use one of the special characters as part of the command (I think this is what you were getting at initially).

Also, was reading /usr/share/bash-completion/bash_completion _quote_readline_by_ref() and it said that compgen needs its arguments quoted/escaped, so, yes, we need to handle bounced.

One last thing, '&' is just as bad as ';' so it should be added to the grep regex. However, really starting to think we need to remove that and add quoting....